src/main/java/com/ita/if103java/ims/dao/impl/WarehouseDaoImpl.java

```diff
@@ -104,20 +115,16 @@ public Warehouse findById(Long id, Long accountId) {
     @Override
     public Warehouse create(Warehouse warehouse) {
         try {
-            GeneratedKeyHolder holder = new GeneratedKeyHolder();
-            jdbcTemplate.update(connection -> createWarehouseStatement(warehouse, connection), holder);
-            warehouse.setId(Optional.ofNullable(holder.getKey())
-                .map(Number::longValue)
-                .orElseThrow(() -> new CRUDException("Error during an warehouse creation: " +
-                    "Autogenerated key is null")));
-
+            Long id = createWithAutogeneratedId(holder ->
+                jdbcTemplate.update(connection -> createWarehouseStatement(warehouse, connection), holder)
+            );
+            warehouse.setId(id);
```

Comment:
If you didn't change the concept and all warehouses regardless of their level should have top warehouse id. Even warehouse with 0 level, then here I would like to add the following:
```
if (warehouse.isTopLevel()) {
  warehouse.setTopLevelWarehouseId(id);
  this.update(warehouse);
}
```